### ISSUE TITLE
Why GAN example use MultiGPUGANTrainer instead of SyncMultiGPUTrainerReplicated?

### ISSUE THREAD

**Issue by yogurfrul** [tensorpack/tensorpack#878]
1.in the example Image2Image, you define gpu flag while the trainer is still GANTrainer which is running on just one gpu,it shoule be as below as BEGAN example:
```
       nr_tower = max(get_num_gpu(), 1)
        if nr_tower == 1:
            trainer = GANTrainer(input, model)
        else:
            trainer = MultiGPUGANTrainer(nr_tower, input, model)

```

2.why you define GANTrainer and MultiGPUGANTrainer instead of  SimpleTrainer and SyncMultiGPUTrainerReplicated? Can I use HorovodTrainer?

3.I can see the MultiGPUGANTrainer do the gradient allreduce using add_n ,why not use ncclAllreduce?

4.all the example of GAN use the build_losses defined in GANModelDesc,while the function description showing that it's standard GAN loss , I think it shoule be cGAN loss with logits_real
```
        Build standard GAN loss and set `self.g_loss` and `self.d_loss`.

        D and G play two-player minimax game with value function V(G,D)

          min_G max _D V(D, G) = IE_{x ~ p_data} [log D(x)] + IE_{z ~ p_fake} [log (1 - D(G(z)))]
```

**Comment by ppwwyyxx** [comment]
1. contributions are welcome
2. No. all trainers in tensorpack are single-cost trainers, as you can see here: https://tensorpack.readthedocs.io/modules/train.html
3. both are ok
4. Would you elaborate?

**Comment by yogurfrul** [comment]
I make a mistake in question 3, the MultiGPUGANTrainer just add and mean of all the loss ,not the variable gradient. I think it's a dangerous trick ,if this is right, why we allreduce the gradient instead of loss?

**Comment by ppwwyyxx** [comment]
Where do you think it is not right?

It's faster to reduce the gradients.

**Comment by yogurfrul** [comment]
you mean that  reduce the loss is equal to reduce the gradient?
And the reduce of gradient is faster than reduce of loss?
Size of gradient  is same with the shape of tranning variable , bigger than the loss,which is just scalar in common model. How can reduce of gradient is faster?

**Comment by ppwwyyxx** [comment]
If you reduce the loss and compute gradients on the reduced loss, tensorflow will copy the gradients among GPUs.
